src/llmq/quorums_blockprocessor.cpp

```diff
@@ -0,0 +1,443 @@
+                    if (jt->second.CountSigners() <= qc.CountSigners()) {
+                        return;
+                    }
+                }
+            }
+        }
+
+        auto members = CLLMQUtils::GetAllQuorumMembers(type, qc.quorumHash);
+
+        if (!qc.Verify(members)) {
+            LOCK(cs_main);
+            LogPrintf("CQuorumBlockProcessor::%s -- commitment for quorum %s:%d is not valid, peer=%d\n", __func__,
+                      qc.quorumHash.ToString(), qc.llmqType, pfrom->id);
+            Misbehaving(pfrom->id, 100);
+            return;
+        }
+
+        LogPrintf("CQuorumBlockProcessor::%s -- received commitment for quorum %s:%d, validMembers=%d, signers=%d, peer=%d\n", __func__,
+                  qc.quorumHash.ToString(), qc.llmqType, qc.CountValidMembers(), qc.CountSigners(), pfrom->id);
+
+        AddMinableCommitment(qc);
+    }
+}
+
+bool CQuorumBlockProcessor::ProcessBlock(const CBlock& block, const CBlockIndex* pindexPrev, CValidationState& state)
+{
+    AssertLockHeld(cs_main);
+
+    bool fDIP0003Active = VersionBitsState(pindexPrev, Params().GetConsensus(), Consensus::DEPLOYMENT_DIP0003, versionbitscache) == THRESHOLD_ACTIVE;
+    if (!fDIP0003Active) {
+        return true;
+    }
+
+    int nHeight = pindexPrev->nHeight + 1;
+
+    std::map<Consensus::LLMQType, CFinalCommitment> qcs;
+    if (!GetCommitmentsFromBlock(block, pindexPrev, qcs, state)) {
+        return false;
+    }
+
+    for (const auto& p : Params().GetConsensus().llmqs) {
+        auto type = p.first;
+
+        uint256 quorumHash = GetQuorumBlockHash(type, pindexPrev);
+
+        if (!quorumHash.IsNull() && IsMiningPhase(type, nHeight)) {
+            if (HasMinedCommitment(type, quorumHash)) {
+                if (qcs.count(type)) {
+                    // If in the current mining phase a previous block already mined a non-null commitment and the new
+                    // block contains another (null or non-null) commitment, the block should be rejected.
+                    return state.DoS(100, false, REJECT_INVALID, "bad-qc-already-mined");
+                }
+            } else {
+                if (!qcs.count(type)) {
+                    bool allowMissingQc = false;
+                    {
+                        // TODO We added the commitments code while DIP3 was already activated on testnet and we want
+                        // to avoid reverting the chain again, as people already had many MNs registered at that time.
+                        // So, we do a simple hardfork here at a fixed height, but only while we're on the original
+                        // DIP3 chain.
+                        // As we need to fork/revert the chain later to re-test all deployment stages of DIP3, we can
+                        // remove all this temporary code later.
+                        LOCK(cs_main);
+                        int oldTestnetDIP3ActivationHeight = 264000;
```

Comment:
I'm assuming these are part of the temporary code, otherwise should probably move to chainparams.

Reply:
Yepp, this is going to vanish when we reset testnet to do the full retest of everything